File: src/logging_learning_cpp/src/logging_levels.cpp
/**
 * @file logging_levels.cpp
 * @brief ② 日志级别控制 —— 运行时动态调整与外部配置
 *
 * 知识点：
 * - 通过参数设置日志级别
 * - 命名日志器的独立级别控制
 * - 日志器层级与继承
 * - 运行时动态切换日志级别
 */

#include <rclcpp/rclcpp.hpp>

class LoggingLevelsNode : public rclcpp::Node
{
public:
  // Jazzy 中 rclcpp::Logger 默认构造函数为 private，
  // 必须在成员初始化列表中初始化，不能先默认构造再赋值
  LoggingLevelsNode() : Node("logging_levels"),
    sensor_logger_(this->get_logger().get_child("sensor")),
    motor_logger_(this->get_logger().get_child("motor"))
  {
    // ================================================================
    // 方式 1：通过参数设置日志级别
    // ================================================================
    this->declare_parameter("log_level", std::string("info"));

    std::string log_level_str = this->get_parameter("log_level").as_string();
    auto level = level_from_string(log_level_str);
    this->get_logger().set_level(level);
    RCLCPP_INFO(this->get_logger(), "Log level set via parameter: %s", log_level_str.c_str());

    // ================================================================
    // 命名日志器 —— 为子模块创建独立日志器
    // ================================================================

    // 子日志器已在初始化列表中创建
    // （层级关系：logging_levels.sensor / logging_levels.motor）

    // 为子日志器设置不同级别
    sensor_logger_.set_level(rclcpp::Logger::Level::Debug);  // 传感器模块看调试日志
    motor_logger_.set_level(rclcpp::Logger::Level::Warn);    // 电机模块只看警告以上

    RCLCPP_INFO(sensor_logger_, "Sensor logger initialized (DEBUG level)");
    RCLCPP_DEBUG(sensor_logger_, "Sensor debug info visible");
    RCLCPP_INFO(motor_logger_, "Motor logger initialized (WARN level, this INFO is invisible)");
    RCLCPP_WARN(motor_logger_, "Motor warning is visible");

    // ================================================================
    // 定时器演示日志级别动态切换
    // ================================================================
    timer_ = this->create_wall_timer(
      std::chrono::seconds(2),
      std::bind(&LoggingLevelsNode::timer_callback, this));
  }

private:
  rclcpp::Logger sensor_logger_;
  rclcpp::Logger motor_logger_;
  rclcpp::TimerBase::SharedPtr timer_;
  int count_ = 0;

  void timer_callback()
  {
    count_++;

    // 每隔 10 次切换一次日志级别
    if (count_ % 10 == 0) {
      auto new_level = (count_ / 10) % 2 == 0
        ? rclcpp::Logger::Level::Debug
        : rclcpp::Logger::Level::Info;
      this->get_logger().set_level(new_level);
      RCLCPP_INFO(this->get_logger(), "Log level switched to %s",
        new_level == rclcpp::Logger::Level::Debug ? "DEBUG" : "INFO");
    }

    RCLCPP_DEBUG(this->get_logger(), "Callback #%d (DEBUG)", count_);
    RCLCPP_INFO(this->get_logger(), "Callback #%d (INFO)", count_);
    RCLCPP_WARN(this->get_logger(), "Callback #%d (WARN)", count_);
  }

  static rclcpp::Logger::Level level_from_string(const std::string & level_str)
  {
    if (level_str == "debug") return rclcpp::Logger::Level::Debug;
    if (level_str == "info")  return rclcpp::Logger::Level::Info;
    if (level_str == "warn")  return rclcpp::Logger::Level::Warn;
    if (level_str == "error") return rclcpp::Logger::Level::Error;
    if (level_str == "fatal") return rclcpp::Logger::Level::Fatal;
    return rclcpp::Logger::Level::Info;  // 默认
  }
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<LoggingLevelsNode>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: src/logging_learning_py/logging_learning_py/logging_levels.py
"""② 日志级别控制 —— 运行时动态调整与命名日志器"""
import rclpy
from rclpy.node import Node
from rclpy.logging import LoggingSeverity


class LoggingLevelsNode(Node):
    """日志级别控制演示节点"""

    def __init__(self):
        super().__init__('logging_levels')

        # ================================================================
        # 通过参数设置日志级别
        # ================================================================
        self.declare_parameter('log_level', 'info')
        log_level_str = self.get_parameter('log_level').get_parameter_value().string_value
        level = self._level_from_string(log_level_str)
        self.get_logger().set_level(level)
        self.get_logger().info(f'Log level set via parameter: {log_level_str}')

        # ================================================================
        # 命名日志器
        # ================================================================
        self._sensor_logger = self.get_logger().get_child('sensor')
        self._motor_logger = self.get_logger().get_child('motor')

        self._sensor_logger.set_level(LoggingSeverity.DEBUG)
        self._motor_logger.set_level(LoggingSeverity.WARN)

        self._sensor_logger.info('Sensor logger initialized (DEBUG level)')
        self._sensor_logger.debug('Sensor debug info visible')
        self._motor_logger.info('Motor INFO invisible (WARN level)')
        self._motor_logger.warning('Motor warning is visible')

        # ================================================================
        # 定时器演示
        # ================================================================
        self._count = 0
        self._timer = self.create_timer(2.0, self._timer_callback)

    def _timer_callback(self):
        self._count += 1

        # 每 10 次切换级别
        if self._count % 10 == 0:
            new_level = (
                LoggingSeverity.DEBUG
                if (self._count // 10) % 2 == 0
                else LoggingSeverity.INFO
            )
            self.get_logger().set_level(new_level)
            level_name = 'DEBUG' if new_level == LoggingSeverity.DEBUG else 'INFO'
            self.get_logger().info(f'Log level switched to {level_name}')

        self.get_logger().debug(f'Callback #{self._count} (DEBUG)')
        self.get_logger().info(f'Callback #{self._count} (INFO)')
        self.get_logger().warning(f'Callback #{self._count} (WARN)')

    @staticmethod
    def _level_from_string(level_str: str):
        """字符串转日志级别"""
        levels = {
            'debug': LoggingSeverity.DEBUG,
            'info': LoggingSeverity.INFO,
            'warn': LoggingSeverity.WARN,
            'error': LoggingSeverity.ERROR,
            'fatal': LoggingSeverity.FATAL,
        }
        return levels.get(level_str.lower(), LoggingSeverity.INFO)


def main(args=None):
    rclpy.init(args=args)
    node = LoggingLevelsNode()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()
